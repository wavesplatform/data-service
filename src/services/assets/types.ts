import { CacheSync, Service, XOR, Asset } from '../../types';
import { RequestWithCursor } from '../_common/pagination';
import { Cursor } from './cursor';

export type AssetsCache = CacheSync<string, AssetDbResponse>;

export type SearchByTicker = {
  ticker: string;
  limit: number;
};

export type SearchFullText = {
  search: string;
  limit: number;
};

export type AssetsSearchRequest = RequestWithCursor<
  XOR<SearchByTicker, SearchFullText>,
  Cursor
>;

export type AssetsService = Service<
  string,
  string[],
  AssetsSearchRequest,
  Asset
>;

export type AssetDbResponse = {
  asset_id: string;
  asset_name: string;
  description: string;
  sender: string;
  issue_height: number;
  total_quantity: string;
  decimals: number;
  reissuable: boolean;
  ticker: string;
  issue_timestamp: Date;
  has_script: boolean;
  min_sponsored_asset_fee: number;
};
