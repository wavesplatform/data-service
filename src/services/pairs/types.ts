import { AssetIdsPair } from '../../types';

export type MgetRequest = {
  pairs: AssetIdsPair[];
  matcher: string;
};
