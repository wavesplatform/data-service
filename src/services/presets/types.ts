import { PgDriver } from 'db/driver';
import { EmitEvent } from 'services/_common/createResolver/types';

export type ServicePresetInitOptions = {
  pg: PgDriver;
  emitEvent: EmitEvent;
};
