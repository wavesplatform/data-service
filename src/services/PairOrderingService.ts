import { Task, of as taskOf } from 'folktale/concurrency/task';
import { Maybe, empty as maybeEmpty } from 'folktale/maybe';
import { AppError } from 'errorHandling';

export interface PairOrderingService {
  getCorrectOrder(
    matcher: string,
    pair: [string, string]
  ): Task<AppError, Maybe<[string, string]>>;
}

export class PairOrderingServiceImpl {
  private constructor(matcherSettings: string[]) {}

  public static dummy(): Task<AppError, PairOrderingService> {
    return taskOf({
      getCorrectOrder() {
        return taskOf<AppError, Maybe<[string, string]>>(maybeEmpty());
      },
    });
  }
}
