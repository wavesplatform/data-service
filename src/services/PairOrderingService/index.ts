import { Task, waitAll } from 'folktale/concurrency/task';
import { Maybe, empty as nothing, of as just } from 'folktale/maybe';
import { InitError } from '../../errorHandling';
import { AssetIdsPair } from '../../types';
import { loadMatcherSettings } from './loadMatcherSettings';

import { zipObj, map } from 'ramda';

import { createOrderPair, TOrderPair } from '@waves/assets-pairs-order';

export interface PairOrderingService {
  isCorrectOrder(matcher: string, pair: AssetIdsPair): Maybe<boolean>;

  getCorrectOrder(matcher: string, pair: [string, string]): Maybe<AssetIdsPair>;
}

export class PairOrderingServiceImpl implements PairOrderingService {
  private orderPair: Record<string, TOrderPair>;

  constructor(matchersSettings: Record<string, string[]>) {
    this.orderPair = map(createOrderPair, matchersSettings);
  }

  public static create(
    matchersSettingsURLs: Record<string, string>
  ): Task<InitError, PairOrderingService> {
    const matcherAddresses = Object.keys(matchersSettingsURLs);

    const matcherSettingsTasks = matcherAddresses.map(ma =>
      loadMatcherSettings(matchersSettingsURLs[ma]).map(s => s.priceAssets)
    );

    return waitAll(matcherSettingsTasks)
      .map(settings => zipObj(matcherAddresses, settings))
      .map(s => new PairOrderingServiceImpl(s));
  }

  getCorrectOrder(
    matcher: string,
    pair: [string, string]
  ): Maybe<AssetIdsPair> {
    if (!this.orderPair[matcher]) return nothing();
    else {
      const correctOrder = this.orderPair[matcher](pair[0], pair[1]);
      return just({
        amountAsset: correctOrder[0],
        priceAsset: correctOrder[1],
      });
    }
  }

  public isCorrectOrder(matcher: string, pair: AssetIdsPair): Maybe<boolean> {
    return this.getCorrectOrder(matcher, [
      pair.amountAsset,
      pair.priceAsset,
    ]).map(p => p.amountAsset === pair.amountAsset);
  }
}
