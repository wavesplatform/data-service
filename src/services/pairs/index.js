import { getByIdPreset } from '../presets/pg/getById';
import { searchPreset } from '../presets/pg/search';
import { pair } from '../../types';

import mget from './mget';
import { inputGet, inputSearch, result as resultSchema } from './schema';
import { transformResult, transformResultSearch } from './transformResult';
import * as sql from './sql';

module.exports = ({ drivers, emitEvent }) => {
  return {
    get: getByIdPreset({
      name: 'pairs.get',
      sql: sql.get,
      inputSchema: inputGet,
      resultSchema,
      transformResult: transformResult,
      resultTypeFactory: pair,
    })({ pg: drivers.pg, emitEvent }),
    mget: mget({
      name: 'pairs.mget',
      sql: sql.mget,
      transformResult: transformResult,
      typeFactory: pair,
    })({ pg: drivers.pg, emitEvent }),
    search: searchPreset({
      name: 'pairs.search',
      sql: sql.search,
      inputSchema: inputSearch,
      resultSchema,
      transformResult: transformResultSearch,
    })({ pg: drivers.pg, emitEvent }),
  };
};
