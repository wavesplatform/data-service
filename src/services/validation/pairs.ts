import { PairOrderingService } from '../PairOrderingService';
import { AssetIdsPair, ServiceMget, Asset } from '../../types';
import {
  Task,
  rejected as taskRejected,
  of as taskOf,
} from 'folktale/concurrency/task';
import { ValidationError, AppError } from '../../errorHandling/index';
import { zip } from 'ramda';

export const validatePairs = (
  assetsMget: ServiceMget<string[], Asset>,
  pairOrderingService: PairOrderingService
) => (matcher: string, pairs: AssetIdsPair[]): Task<AppError, void> => {
  // correct order
  const incorrectPairs = pairs.filter(
    p =>
      !pairOrderingService.isCorrectOrder(matcher, p).matchWith({
        Just: ({ value }) => value,
        Nothing: () => true,
      })
  );

  if (incorrectPairs.length)
    return taskRejected(
      new ValidationError('Wrong assets order in provided pair(s)', {
        pairs: incorrectPairs,
      })
    );

  // all assets exist
  const assetIdsSet: Set<string> = new Set();
  pairs.forEach(p => {
    assetIdsSet.add(p.amountAsset);
    assetIdsSet.add(p.priceAsset);
  });
  const assetIds = Array.from(assetIdsSet);

  return assetsMget.mget(assetIds).chain(assets => {
    const nonExistingIds = zip(assetIds, assets.data)
      .filter(x => x[1].data === null)
      .map(x => x[0]);

    if (!nonExistingIds.length) {
      return taskOf(undefined);
    } else {
      return taskRejected(
        new ValidationError('Assets do not exist in the blockchain', {
          assets: nonExistingIds,
        })
      );
    }
  });
};
