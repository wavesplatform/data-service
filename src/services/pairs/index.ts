import { identity } from 'ramda';
import { of as taskOf, Task } from 'folktale/concurrency/task';
import { of as just, Maybe } from 'folktale/maybe';

import { forEach, isEmpty } from '../../utils/fp/maybeOps';
import { tap } from '../../utils/tap';

import { withStatementTimeout } from '../../db/driver';
import { AppError } from '../../errorHandling';
import {
  pair,
  PairInfo,
  Pair,
  List,
  Service,
  CacheSync,
  AssetIdsPair,
} from '../../types';
import { CommonServiceDependencies } from '..';

// resolver creation and presets
import {
  get as createGetResolver,
  mget as createMgetResolver,
} from '../_common/createResolver';
import { getData as getByIdPg } from '../presets/pg/getById/pg';
import { validateInput, validateResult } from '../presets/validation';
import { searchPreset } from '../presets/pg/search';

// service logic
import { matchRequestResult } from './matchRequestResult';
import { mgetPairsPg } from './mgetPairsPg';
import {
  inputGet,
  inputMget,
  inputSearch,
  result as resultSchema,
} from './schema';
import { transformResults as transformResultMget } from './transformResults';
import {
  PairDbResponse,
  transformResult,
  transformResultSearch,
} from './transformResult';
import * as sql from './sql';

export type PairsGetRequest = {
  pair: AssetIdsPair;
  matcher: string;
};

export type PairsMgetRequest = { pairs: AssetIdsPair[]; matcher: string };

export type SearchCommonRequest = {
  matcher: string;
  limit: number;
};

export type SearchWithMatchExactly = SearchCommonRequest & {
  match_exactly?: boolean[];
};

export type SearchByAssetRequest = SearchWithMatchExactly & {
  search_by_asset: string;
};

export type SearchByAssetsRequest = SearchWithMatchExactly & {
  search_by_assets: [string, string];
};

export type PairsSearchRequest =
  | SearchCommonRequest
  | SearchByAssetRequest
  | SearchByAssetsRequest;

export type PairsService = Service<
  PairsGetRequest,
  PairsMgetRequest,
  PairsSearchRequest,
  Pair
>;

export { create as createCache } from './cache';

export default ({
  drivers,
  emitEvent,
  timeouts,
  validatePairs,
  cache,
}: CommonServiceDependencies & {
  validatePairs: (
    matcher: string,
    pairs: AssetIdsPair[]
  ) => Task<AppError, void>;
  cache: CacheSync<PairsGetRequest, PairDbResponse>;
}): PairsService => {
  const SERVICE_NAME = {
    GET: 'pairs.get',
    MGET: 'pairs.mget',
    SEARCH: 'pairs.search',
  };

  const get = createGetResolver<
    PairsGetRequest,
    PairsGetRequest,
    PairDbResponse,
    Pair
  >({
    transformInput: identity,
    validateInput: value =>
      validateInput<PairsGetRequest>(inputGet, SERVICE_NAME.GET)(value).chain(
        v => validatePairs(v.matcher, [v.pair]).map(() => v)
      ),

    // cache first
    getData: req =>
      cache.get(req).matchWith({
        Just: ({ value }) => taskOf(just(value)),
        Nothing: () =>
          getByIdPg<PairDbResponse, PairsGetRequest>({
            name: SERVICE_NAME.GET,
            sql: sql.get,
            pg: withStatementTimeout(
              drivers.pg,
              timeouts.get,
              timeouts.default
            ),
          })(req).map(
            tap(maybeResp => forEach(x => cache.set(req, x), maybeResp))
          ),
      }),
    validateResult: validateResult(resultSchema, SERVICE_NAME.GET),
    transformResult: res =>
      res.map(res =>
        pair(transformResult(res), {
          amountAsset: res.amount_asset_id,
          priceAsset: res.price_asset_id,
        })
      ),
    emitEvent,
  });

  const mget = createMgetResolver<
    PairsMgetRequest,
    PairsMgetRequest,
    PairDbResponse,
    List<Pair>
  >({
    transformInput: identity,
    validateInput: value =>
      validateInput<PairsMgetRequest>(inputMget, SERVICE_NAME.MGET)(
        value
      ).chain(v => validatePairs(v.matcher, v.pairs).map(() => v)),
    getData: request => {
      let results: Array<Maybe<PairDbResponse>> = request.pairs.map(p =>
        cache.get({
          pair: p,
          matcher: request.matcher,
        })
      );

      const notCachedIndexes = results.reduce<number[]>((acc, x, i) => {
        if (isEmpty(x)) acc.push(i);
        return acc;
      }, []);

      const notCachedPairs = notCachedIndexes.map(i => request.pairs[i]);

      return mgetPairsPg({
        name: SERVICE_NAME.MGET,
        sql: sql.mget,
        matchRequestResult,
        pg: withStatementTimeout(drivers.pg, timeouts.mget, timeouts.default),
      })({
        pairs: notCachedPairs,
        matcher: request.matcher,
      }).map(pairsFromDb => {
        pairsFromDb.forEach((pair, index) =>
          forEach(p => {
            results[notCachedIndexes[index]] = pair;
            cache.set(
              {
                matcher: request.matcher,
                pair: notCachedPairs[index],
              },
              p
            );
          }, pair)
        );
        return results;
      });
    },
    validateResult: validateResult(resultSchema, SERVICE_NAME.MGET),
    transformResult: transformResultMget,
    emitEvent,
  });

  const search = searchPreset<
    PairsSearchRequest,
    PairDbResponse,
    PairInfo,
    List<Pair>
  >({
    name: SERVICE_NAME.SEARCH,
    sql: sql.search,
    inputSchema: inputSearch,
    resultSchema,
    transformResult: transformResultSearch,
  })({
    pg: withStatementTimeout(drivers.pg, timeouts.search, timeouts.default),
    emitEvent,
  });

  return {
    get,
    mget,
    search,
  };
};
