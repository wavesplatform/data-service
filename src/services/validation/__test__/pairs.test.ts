import { validatePairs } from '../pairs';

import { PairOrderingServiceImpl } from '../../PairOrderingService';
import { asset, AssetInfo, ServiceMget, Asset, list } from '../../../types';
import { of as taskOf } from 'folktale/concurrency/task';

describe('Pairs validation', () => {
  const MATCHER = 'matcher';
  const WAVES = 'WAVES';
  const BTC = 'BTC';

  const pairOrderingService = new PairOrderingServiceImpl({
    [MATCHER]: [BTC, WAVES],
  });

  const assetsMget: ServiceMget<string[], Asset> = {
    mget: (assetIds: string[]) =>
      taskOf(
        list(
          assetIds.map(aid => {
            switch (aid) {
              case BTC:
              case WAVES:
                return asset({} as AssetInfo);
              default:
                return asset(null);
            }
          })
        )
      ),
  };

  const validate = validatePairs(assetsMget, pairOrderingService);

  describe('asset order validation', () => {
    it('known matcher, right order, pass', () =>
      expect(
        validate(MATCHER, [{ amountAsset: WAVES, priceAsset: BTC }])
          .run()
          .promise()
      ).resolves.not.toThrow());

    it('unknown matcher, existing assets, pass', () =>
      expect(
        validate('', [{ amountAsset: WAVES, priceAsset: BTC }])
          .run()
          .promise()
      ).resolves.not.toThrow());

    it('known matcher, wrong order, fail', () =>
      expect(
        validate(MATCHER, [{ amountAsset: BTC, priceAsset: WAVES }])
          .run()
          .promise()
      ).rejects.toMatchSnapshot());
  });

  describe('assets existence validation', () => {
    it('non-existing assets, right order, fail', () =>
      expect(
        validate(MATCHER, [{ amountAsset: 'ASSET1', priceAsset: BTC }])
          .run()
          .promise()
      ).rejects.toMatchSnapshot());

    it('non-existing assets, wrong order, fail with ordering error', () =>
      expect(
        validate(MATCHER, [{ amountAsset: 'ASSET1', priceAsset: 'ASSET2' }])
          .run()
          .promise()
      ).rejects.toMatchSnapshot());
  });
});
