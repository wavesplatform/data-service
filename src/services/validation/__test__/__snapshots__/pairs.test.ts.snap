// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pairs validation asset order validation known matcher, wrong order, fail 1`] = `
ValidationError {
  "error": [Error: Wrong assets order in provided pair(s)],
  "meta": Object {
    "pairs": Array [
      Object {
        "amountAsset": "BTC",
        "priceAsset": "WAVES",
      },
    ],
  },
  "type": "Validation",
}
`;

exports[`Pairs validation assets existence validation non-existing assets, right order, fail 1`] = `
ValidationError {
  "error": [Error: Assets do not exist in the blockchain],
  "meta": Object {
    "assets": Array [
      "ASSET1",
    ],
  },
  "type": "Validation",
}
`;

exports[`Pairs validation assets existence validation non-existing assets, wrong order, fail with ordering error 1`] = `
ValidationError {
  "error": [Error: Wrong assets order in provided pair(s)],
  "meta": Object {
    "pairs": Array [
      Object {
        "amountAsset": "ASSET1",
        "priceAsset": "ASSET2",
      },
    ],
  },
  "type": "Validation",
}
`;
